src/x_system/x_list.1
.TH X_LIST 1 "Feb 27 2014" "GMT 4.5.13 (SVN)" "Generic Mapping Tools"
.SH NAME
\fBx_list\fP \- Extract Cross-Over Information from the Data Base.
.SH SYNOPSIS
\fBx_list\fP [ \fIleg1\fP ] [ \fIleg2\fP ] [ \fB\-\fP\fIdataflags\fP ] 
[ \fB\-R\fP\fIwest/east/south/north\fP ] 
[ \fB\-X\fP\fIxbase\fP ] [ \fB\-H\fP ] [ \fB\-I\fP ] 
[ \fB\-V\fP ] [ \fB\-C\fP\fIcorrfile\fP ] [ \fB\-Z\fP\fIignorefile\fP ]
.SH DESCRIPTION
\fBx_list\fP will allow the user to extract information from the xx_base.b and xx_legs.b files.
Several options can be passed on the command line in order to specify which parameters to
extract. They are:
.TP
\fIleg1 leg2\fP
If two legs are specified, the cross-overs they have in common are returned.  If one leg is
specified, all the external cross-overs involving this leg are returned, OR only its internal
cross-overs if \-I is selected.  If no legs are given, all the external (or internal with \-I)
cross-overs are returned.  This program is useful if one wants to look at, say, gravity COEs versus
time from port for a particular leg, etc.
.TP
\fIdataflags\fP
This is a string of any combination of the following:
.br
	\fBl\fP means list both legnames.
.br
	\fBt\fP means list time.
.br
	\fBx\fP means list longitude.
.br
	\fBy\fP means list latitude.
.br
	\fBg\fP means list gravity COEs.
.br
	\fBm\fP means list magnetics COEs.
.br
	\fBb\fP means list bathymetry COEs.
.br
	\fBG\fP means list average gravity at cross-over.
.br
	\fBM\fP means list average magnetic anomaly at cross-over.
.br
	\fBB\fP means list average bathymetry at cross-over.
.br
	\fBh\fP means list heading of ship at cross-over.
.br
The components are written out in the order they appear in \fIdataflags\fP. The default output is
\fB\-txygmbGMBhl\fP.  When internal COEs are desired, the time reported is the elapsed time since
the ship first occupied the cross-over point. For external COEs the time means time from the start
of the year in seconds.
.TP
\fB\-R\fP
Only return cross-overs inside the specified region \fIwest, east, south, \fPand \fInorth\fP.  [Default is world].
.TP
\fB\-X\fP
Indicate alternate \fIxx_base.b\fP file.
.TP
\fB\-H\fP
Issue one header record on output.
.TP
\fB\-I\fP
Report on internal COEs [Default is external COEs].
.TP
\fB\-C\fP
Apply cross-over correction to the data (i.e., reports the cross-over value after the best-fitting
corrections have been applied to both legs). If no \fIcorrfile\fP is given, the default correction file
is assumed.
.TP
\fB\-Z\fP
Ignore those legs that appear in the \fIignorefile\fP.
.TP
\fB\-V\fP
Selects verbose mode, which will send progress reports to stderr [Default runs "silently"].
.SH SEE ALSO
.IR GMT (1),
.IR x_system (1)
.SH REFERENCES
Wessel, P. XOVER: A Cross-over Error Detector for Track Data,
\fIComputers & Geosciences, 15\fP, 333-346.


ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 12925 2014-02-20 22:02:24Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
201x-yy-zz	pwessel
	GMT 5.1.1 finalized and released.
-----------------------------------------------------------------------------------------
2014-02-20	pwessel
	* grdcontour.c:		The -Zscale/offset option was applied incorrectly due to a
				change in the underlying scaling function.
	* grdvolume.c:		---"---
2014-02-10	pwessel
	*gmt2kml.c:		Make sure the default style IDs are unique for each process.

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.
2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		We consider the change in SDIST, PDIST to return degrees instead
				of km to be a bug as it makes it difficult to handle geodesics
				for which degrees make no sense.  Thus we have consolidated all
				the ?DIST operators to give km if geographic data are involved.
				Example 23 has been adjusted accordingly, as has several test
				scripts.  Please check your own scripts for SDIST.

2014-01-09	pwessel
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
2014-01-09	remko
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
	* grdmath.c:		LDIST and PDIST will again produce km when using geodesics.

2014-01-01	pwessel
	* x2sys/x2sys.c:	Mishandled the assignment of segment number for each record.
	* gmt_support.c:	Allow GMT_crossover to deal with global wrapped data.

2013-12-31	pwessel
	* gmt2kml.c:		Now using GMT_Put_Record instead of old stdout printing.
				This addresses issue #473.

2013-12-10	pwessel
	* grdvector.c:		Vector direction did not adjust when negative scales were
				used in -JX or -Jx.
	* psxy[z].c		---"---

2013-12-08	remko
	* psimage.c:		Now properly plots transparent images and makes -Gt work.

2013-11-26	pwessel
	* gmt_io.c:		GMT would crash on a blank line with no newline at end at all.
	* pstext.c:		pstext did not process GMT4-style input properly.

2013-11-26	jluis
	* potential/gravfft.c:	Let -E options default to 3.
				Let users select the -N+l (leave trend alone). That is, do not overwrite with default mode.
				Let the default detrend be remove MID point.
				Explicitly set -N+a in gravfft command to avoid failures due to coming changes in gravfft.
2013-11-25	pwessel
	* grdlandmask.c:	Failed to fill partial bins with no coastlines when the region
				crossed Greenwich [issue #451].

2013-11-21	pwessel
	* pscontour.c:		Let pscontour output contours to file [issue #448].
	* pscoast.c:		Prevent users from not giving proper projection [issue #449].

2013-11-19	jluis
	* mgd77/mgd77magref.c:	Needed to reset the t_col variable when constant altitude [issue #446].

2013-11-19	pwessel
	* gmt_sph.c:		Fix invalid read in ssrfpack_grid due to par array being of length 2 rather than 3.

2013-11-18	pwessel
	* gmt_grdio.c:		Place the old auto-detect of 4-byte binary float vs int under compatibility mode for backwardness.

2013-11-16	pwessel
	* grdvector.c:		Address shortcomings of geovectors when wrapped and clipped.

2013-11-15	pwessel
	* geovectors.c:		Fix geovector when shrinking is selected.

2013-11-13	pwessel
	* gmt_api.c:		Fix API issues pointed out by Ian.

2013-11-13	fwobbe
	* GMT.app:		Inherit user shell in GMT.app.

2013-11-12	pwessel
	* gmt_init.c:		grdmath could corrupt the gmt.history file [issue # 438].
 	
2013-11-10	pwessel
	* gmt_support.c:	Fix symbol-size issue in psxy -Sc [issue #435].
	
2013-11-09	pwessel
	* grdvector.c:		Did not convert direction to azimuths for geo vectors.

2013-11-08	pwessel
	* grdvector.c:		Fix grdvector -I problem [issue #434].
	* ps2raster.c:		Under Mavericks, strcpy would exit if the to and from pointers
-----------------------------------------------------------------------------------------
2013-11-05	pwessel
	GMT 5.1.0 finalized and released.
-----------------------------------------------------------------------------------------

2013-11-01	remko
	* example_10.sh:	Changed input to number of languages; multi-color columns

2013-10-31	pwessel
	* gmt_dcw.c:		Added option to specify code as =<continent> which will
				automatically be expanded to all the countries in that
				continent.  E.g., -F=AF+gred will paint Africa read.
				Fixed memory leaks.

2013-10-23	pwessel
	* gmtinfo.c:		The new name of minmax.c, the latter will be honored
				when GMT compatibility mode is 4.
	* grdmath.c:		Fixed porting bugs for CURV and D2DY2; added test ops.sh

2013-10-22	pwessel
	* minmax.c:		Added -D to adjust region found with -I
	* makecpt.c:		Better reporting of gaps or overlaps in CPTs

2013-10-05	pwessel
	* greenspline.c:	Needed -f in order to select -f0T input, plus it made assumptions
				about lon,lat despite not being selected.

2013-10-04	pwessel
	* gmt.history:		This replaces the old .gmtcommands file name.

2013-09-28	remko
	* gmt_init.c:		Moved origin of JD by one day (24 Nov instead of 25 Nov).
	* ConfigUserTemplate.cmake:	Added option GMT_DOCS_DEPEND_ON_GMT

2013-09-27	pwessel
	* share/cpt:		Added 10 new CPTs, mostly for bathymetry and topography.

2013-09-14	pwessel

src/x_system/x_list.c
/*
 *	$Id: x_list.c 9685 2012-02-27 20:18:12Z pwessel $
 *
 * XLIST produces ASCII listings of cross-over info. The xx_base.b-file
 * contains longitude(x), latitude(y), time1(t1), time2(t2),  heading1(h),
 * heading2(h), crossover-values for gravity(g), magnetics(m), and
 * bathymetry(b), plus the average data value at the cross-over point for
 * gravity(G), magnetics(M), and bathymetry(B), and the user may extract any 
 * combination of these 10 parameters + the legnames + time (t1 for external,
 * abs(t2-t1) for internal) by using the option -txyhgmbGMB. The sequence in
 * which the flag characters appear determines the sequence in which the
 * parameters will be printed out.  If no options is specified, the default
 * is -txygmbGMBhl.
 * To select data inside an area, use the -W -E -S -N options:
 *	E.g. all data between 10 and 30 degree East: -W10. -E30.
 * To create a Scientist Helper ASCII Table format file, use -H.
 * 2, 1, or no legnames may be passed on the command line. If two is
 * passed, the external cross-overs generated between these legs are
 * returned. If one is passed, all the external cross-overs involving
 * this legs are returned (or internal if -I is used). No legs means
 * that all external (or internal) cross-overs will be returned.
 *
 * Author:	Paul Wessel
 * Date:	21-JAN-1988
 * Revised:	18-FEB-1989
 *		26-JUN-1989	PW: Print out NaN if no xover. Also -gmb will now
 *				    work even if 1-2 of the xvalues are NaN
 *		06-MAR-2000	PW: POSIX
 *
 */
 
#include "gmt.h"
#include "x_system.h"

#define CORRFILE "/usr/lib/gmt/xx_corrections.b"
#define MAXLEGS 10000

struct CORR **bin;

size_t binsize = sizeof(struct CORR);
GMT_LONG nlegs = 0;
GMT_LONG nbadlegs=0;

char badlegs[BUFSIZ][10];

GMT_LONG get_id (char *name);
GMT_LONG findleg (char *name);

int main (int argc, char *argv[])
{
	GMT_LONG i, j, id1, id2, time, no[12];
	GMT_LONG nval = 0, ok, n_x, nlegs_in = 0, strike, heading;
	char error, g, m, b, SH_format = FALSE, once = FALSE;
	char corrfile[80], correct = FALSE, inter = FALSE, leg[2][10];
	char internal, legchoice = FALSE, header[REC_SIZE], shift_ok, skip, swap;
	char lega[10], legb[10], verbose = FALSE, line[BUFSIZ];
	double lat, lon , xgrv, xmag, xtop, grv, mag, top, t1, t2;
	double west = 0., east = 360., south = -90., north = 90.;
	FILE *fp = NULL, *fpc = NULL, *fpi = NULL;

	struct XOVERS crossover;

	error = FALSE;
	g = m = b = FALSE;

	/* Check and interpret the command line arguments */

	for (i =1; !error && i < argc; i++) {
		if (argv[i][0] == '-') {
			for (j = 1; argv[i][j]; j++) {
				switch(argv[i][j]) {
					case 'g':		/* Gravity xover is requested */
						no[nval++] = 3;
						g = TRUE;
						break;
					case 'm':		/* Magnetics xover is requested */
						no[nval++] = 4;
						m = TRUE;
						break;
					case 'b':		/* Bathymetry xover is requested */
						no[nval++] = 5;
						b = TRUE;
						break;
					case 'G':		/* Gravity is requested */
						no[nval++] = 6;
						g = TRUE;
						break;
					case 'M':		/* Magnetics is requested */
						no[nval++] = 7;
						m = TRUE;
						break;
					case 'B':		/* Bathymetry is requested */
						no[nval++] = 8;
						b = TRUE;
						break;
					case 'x':		/* Longitude is requested */
						no[nval++] = 1;
						break;
					case 'y':		/* Latitude is requested */
						no[nval++] = 2;
						break;
					case 't':		/* Time (in sec) is requested */
						no[nval++] = 0;
						break;
					case 'h':		/*  Heading is requested */
						no[nval++] = 9;
						break;
					case 'l':		/* Legnames are requested */
						no[nval++] = 10;
						no[nval++] = 11;
						break;
					case 'X':
						fp = fopen (&argv[i][j+1], "rb");
						while (argv[i][j]) j++;
						j--;
						break;
					case 'Z':		/* List of legs to ignore */
						fpi = fopen (&argv[i][j+1], "r");
						while (argv[i][j]) j++;
						j--;
						break;
					case 'R':		/* Region specified */
						sscanf (&argv[i][2], "%lf/%lf/%lf/%lf", &west, &east, &south, &north);
						while (argv[i][j]) j++;
						j--;
						break;
					case 'H':		/* Output using Scientist Helper format */
						SH_format = TRUE;
						break;
					case 'V':
						verbose = TRUE;
						break;
					case 'I':		/* Internal Crossovers only */
						inter = TRUE;
						break;
					case 'C':
						if (argv[i][2] == 0)	/* Use default corrfile */
							strcpy (corrfile, CORRFILE);
						else
							strcpy (corrfile, &argv[i][2]);
						correct = TRUE;
						while (argv[i][j]) j++;
	 					j--;
						break;
					default:		/* Options not recognized */
						error = TRUE;
						break;
				}
			}
		}
		else {
			if (nlegs_in < 2) {
				strcpy (leg[nlegs_in], argv[i]);
				nlegs_in++;
				legchoice = TRUE;
			}
			else
				error = TRUE;
		}
	}

	/* Check that the options selected are mutually consistent */

	if (east < west || south > north) error = TRUE;
	if (west < 0.0) {
		west += 360.0;
		east += 360.0;
	}
	if (nlegs_in != 1 && inter) error = TRUE;
	if (nval > 12) error = TRUE;
	if (nlegs_in == 2 && !strcmp(leg[0],leg[1])) {
		inter = TRUE;
		nlegs_in = 1;
	}
	else if (nlegs_in == 2) {
		inter = FALSE;
		if (strcmp(leg[0], leg[1]) > 0) {	/* Not in alphabetical order */
			strcpy (lega, leg[0]);
			strcpy (leg[0], leg[1]);
			strcpy (leg[1], lega);
		}
	}
	if (argc == 1 || error) {	/* Display usage */
		fprintf(stderr,"xlist - Extract information from xover databases\n\n");
		fprintf(stderr,"usage: xlist [leg1] [leg2] [-<dataflags>] [%s]\n", GMT_Rgeo_OPT);
		fprintf(stderr,"	[-X<xbase>] [-H] [-I] [-V] [-C<corrfile> -Z<ignorefile>]\n\n");
		fprintf(stderr,"	Dataflags:\n");
		fprintf(stderr,"	  l means legnames\n");
		fprintf(stderr,"	  t means list time (delta time for internal COEs)\n");
		fprintf(stderr,"	  x means list longitude\n");
		fprintf(stderr,"	  y means list latitude\n");
		fprintf(stderr,"	  g means list gravity cross-over\n");
		fprintf(stderr,"	  m means list magnetics cross-over\n");
		fprintf(stderr,"	  b means list bathymetry cross-over\n");
		fprintf(stderr,"	  G means list average gravity\n");
		fprintf(stderr,"	  M means list average magnetics\n");
		fprintf(stderr,"	  B means list average bathymetry\n");
		fprintf(stderr,"	  h means list heading\n\n");
		fprintf(stderr,"	-R let only data inside the region pass through. [Default is world]\n");
		fprintf(stderr,"	-H will write 1 header record\n");
		fprintf(stderr,"	-I means return internal cross-overs only [Default is external]\n");
		fprintf(stderr,"	-C<file> applies crossover corrections to the data. If no file name\n");
		fprintf(stderr,"	   is given, the default correction file is assumed\n");
		fprintf(stderr,"	-Z will ignore those legs that appear in the ignorefile\n");
		fprintf(stderr,"	-V means verbose\n");
		fprintf(stderr,"	Default is -txygmbGMBhl\n");
		fprintf(stderr,"	The data is written out in the sequence specified.\n");
		exit (EXIT_FAILURE);
	}

	if (correct) {	/* Read correction table */
		if ((fpc = fopen (corrfile,"rb")) == NULL) {
			fprintf(stderr, "Could not read correction file %s\n", corrfile);
			exit (EXIT_FAILURE);
		}
		bin = (struct CORR **) malloc(sizeof(struct CORR *)*MAXLEGS);
		i = 0;
		bin[i] = (struct CORR *) malloc(binsize);
		while (fread((void *)bin[i], binsize, (size_t)1, fpc) == (size_t)1) {
			i++;
			bin[i] = (struct CORR *) malloc(binsize);
		}
		fclose(fpc);
		nlegs = i;
	}

	/* Read the ignore-legs file if needed */
	if (fpi != NULL) {
		while (fgets (line, BUFSIZ, fpi)) {
			sscanf(line, "%s", badlegs[nbadlegs]);
			nbadlegs++;
		}
		fclose(fpi);
	}


	/* Sort the  order in which the parameters appear */

	if (nval == 0) {		/* Nothing selected, default used */
		g = m = b = TRUE;	/* No data was specified so default is used */
		for (i = 0; i < 12; i++)
			no[i] = i;
		nval = 12;
	}	 

	if (fp == NULL && (fp = fopen ("xx_base.b","rb")) == NULL) {
		fprintf(stderr,"Could not open xx_base.b\n");
		exit (EXIT_FAILURE);
	}

	/* Read first record of file containing n_records */

	ok = fread ((void *)header, (size_t)REC_SIZE, (size_t)1, fp);
	ok = fread ((void *)header, (size_t)REC_SIZE, (size_t)1, fp);


	if (SH_format) {
		/* Write out header record */
		for (i = 0; i < nval; i++) {
			switch(no[i]) {
				case 1:	/* Print out time header */
					printf("Time(s)");
					break;
				case 2:	/* Print out latitude */
					printf("lat");
					break;
				case 3:	/* Print out longitude */
					printf ("lon");
					break;
				case 4:	/* Print out gravity cross-overs */
		 			printf ("xfaa");
		 			break;
				case 5:	/* Print out magnetics cross-overs */
		 			printf ("xmag");
					break;
				case 6:	/* Print out bathymetry cross-overs */
					printf ("xtopo");
					break;
				case 7:	/* Print out gravity */
		 			printf ("faa");
		 			break;
				case 8:	/* Print out magnetics */
		 			printf ("mag");
					break;
				case 9:	/* Print out bathymetry */
					printf ("topo");
					break;
				case 10: /* Print out heading */
					printf ("Azimuth");
					break;
				case 11: /* Print out leg_1 */
					printf ("Leg_1");
					break;
				case 12: /* Print out leg_2 */
					printf ("Leg_2");
					break;
			}
			if ((i+1) < nval)
				printf ("%s", gmtdefs.field_delimiter);
			else
				printf ("\n");
		}
	}

	/* Start reading data from file */

	once = (inter && legchoice) || (nlegs_in == 2);

	while (ok) {
		sscanf(header, "%s %s %" GMT_LL "d",lega, legb, &n_x);
		internal = !strcmp(lega, legb);
		skip = FALSE;
		if ((inter && !internal) || (!inter && internal))
			skip = TRUE;
		else if (internal)
			skip = strcmp(leg[0], lega);
		else {	/* External */
			if (nlegs_in == 1)
				skip = (strcmp(leg[0], lega) && strcmp(leg[0], legb));
			else if (nlegs_in == 2)
				skip = (strcmp(leg[0], lega) || strcmp(leg[1], legb));
		}
		if (!skip && nbadlegs > 0 && (findleg (lega) || findleg(legb)))	/* Skip bad leg */
			skip = TRUE;
		if (skip)
			fseek (fp, (long int)n_x*REC_SIZE, SEEK_CUR);
		else {
			swap = (!internal && nlegs_in == 1 && !strcmp(lega,legb)) ? TRUE : FALSE;
			id1 = get_id(lega);
			id2 = (internal) ? id1 : get_id(legb);
			shift_ok = (id1 >=0 && id2 >= 0);
			if (verbose) fprintf (stderr, "%s %s nx=%ld\n", lega, legb, n_x);
			for (j = 0; j < n_x; j++) {
				if ((ok = fread ((void *)&crossover, (size_t)REC_SIZE, (size_t)1, fp)) != (size_t)1) {
					fprintf (stderr, "Read error on xx_base.b\n");
					exit (EXIT_FAILURE);
				}
				strike = 0;
				if (!g || (g && crossover.x_val[0] == NODATA)) strike++;
				if (!m || (m && crossover.x_val[1] == NODATA)) strike++;
				if (!b || (b && crossover.x_val[2] == NODATA)) strike++;
				if (strike == 3) continue;
				lat = (double) crossover.lat*0.000001;
				lon = (double) crossover.lon*0.000001;

				/* Check is lat/lon is outside specified area */

				if (lat < south || lat > north) continue;
				while (lon < west) lon += 360.0;
				if (lon > east) continue;
				t1 = crossover.xtime[0] * 0.1;	/* Since xover time was stored as sec * 10 */
				t2 = crossover.xtime[1] * 0.1;
				for (i = 0; i < nval; i++) {
					switch(no[i]) {
						case 0:	/* Print out time */
							if (internal)
								time = (GMT_LONG)fabs(t2-t1);
							else
								time = (GMT_LONG)((swap) ? t2 : t1);
							printf ("%ld", time);
							break;
						case 1:	/* Print out longitude */
							if (lon > 360.0) lon -= 360.0;
							printf ("%.6g", lon);
							break;
						case 2:	/* Print out latitude */
							printf ("%.6g",lat);
							break;
						case 3:	/* Print out gravity cross-over */
							if (crossover.x_val[0] == NODATA)
								printf ("NaN");
							else {
								xgrv = crossover.x_val[0];
								if (shift_ok) xgrv -= bin[id1]->dc_shift_gmt[0] + bin[id1]->drift_rate_gmt[0] * t1 
								- bin[id2]->dc_shift_gmt[0] - bin[id2]->drift_rate_gmt[0] * t2;
								if (swap) xgrv = -xgrv;
		 						printf ("%.2g", xgrv);
		 					}
		 					break;
						case 4:	/* Print out magnetics cross-over */
							if (crossover.x_val[1] == NODATA)
								printf ("NaN");
							else {
								xmag = crossover.x_val[1];
								if (shift_ok) xmag -= bin[id1]->dc_shift_gmt[1] + bin[id1]->drift_rate_gmt[1] * t1
								- bin[id2]->dc_shift_gmt[1] - bin[id2]->drift_rate_gmt[1] * t2;
								if (swap) xmag = -xmag;
		 						printf ("%ld", (GMT_LONG) xmag);
		 					}
							break;
						case 5:	/* Print out bathymetry cross-over */
							if (crossover.x_val[2] == NODATA)
								printf ("NaN");
							else {
								xtop = crossover.x_val[2];
								if (shift_ok) xtop -= bin[id1]->dc_shift_gmt[2] + bin[id1]->drift_rate_gmt[2] * t1
								- bin[id2]->dc_shift_gmt[2] - bin[id2]->drift_rate_gmt[2] * t2;
								if (swap) xtop = -xtop;
								printf ("%ld", (GMT_LONG) xtop);
							}
							break;
						case 6:	/* Print out gravity */
							if (crossover.gmt[0] == NODATA)
								printf ("NaN");
							else {
								grv = crossover.gmt[0];
								if (shift_ok) grv -= bin[id1]->dc_shift_gmt[0] + bin[id1]->drift_rate_gmt[0] * t1 
								- bin[id2]->dc_shift_gmt[0] - bin[id2]->drift_rate_gmt[0] * t2;
		 						printf ("%.2g", grv);
		 					}
		 					break;
						case 7:	/* Print out magnetics */
							if (crossover.gmt[1] == NODATA)
								printf ("NaN");
							else {
								mag = crossover.gmt[1];
								if (shift_ok) mag -= bin[id1]->dc_shift_gmt[1] + bin[id1]->drift_rate_gmt[1] * t1
								- bin[id2]->dc_shift_gmt[1] - bin[id2]->drift_rate_gmt[1] * t2;
		 						printf ("%ld", (GMT_LONG) mag);
		 					}
							break;
						case 8:	/* Print out bathymetry */
							if (crossover.gmt[2] == NODATA)
								printf ("NaN");
							else {
								top = crossover.gmt[2];
								if (shift_ok) top -= bin[id1]->dc_shift_gmt[2] + bin[id1]->drift_rate_gmt[2] * t1
								- bin[id2]->dc_shift_gmt[2] - bin[id2]->drift_rate_gmt[2] * t2;
								printf ("%ld", (GMT_LONG) top);
							}
							break;
						case 9: /* Print out heading */
							heading = (swap) ? crossover.xhead[1] : crossover.xhead[0];
							printf ("%ld", heading);
							break;
						case 10: /* Print out leg_1 */
							(swap) ? printf ("%s", legb) : printf ("%s", lega);
							break;
						case 11: /* Print out leg_1 */
							(swap) ? printf ("%s", lega) : printf ("%s", legb);
							break;
					}
					if ((i+1) < nval)
						printf ("%s", gmtdefs.field_delimiter);
					else
						printf ("\n");
				}
          		}
          		if (once) {	/* done */
          			fclose (fp);
          			exit (EXIT_FAILURE);
          		}
          	}
		ok = fread ((void *)header, (size_t)REC_SIZE, (size_t)1, fp);
	}
	fclose (fp);

	exit (EXIT_SUCCESS);

}

GMT_LONG get_id (char *name)
{
	GMT_LONG left, right, mid, cmp;

	left = 0;
	right = nlegs-1;
	while (left <= right) {
		mid = (left + right)/2;
		cmp = strcmp(name, bin[mid]->name);
		if (cmp < 0)
			right = mid-1;
		else if (cmp > 0)
			left = mid+1;
		else
			return (mid);
	}
	return (-1);
}

GMT_LONG findleg (char *name)
{
	GMT_LONG left, right, mid, cmp;

	left = 0;
	right = nbadlegs-1;
	while (left <= right) {
		mid = (left + right)/2;
		cmp = strcmp(name, badlegs[mid]);
		if (cmp < 0)
			right = mid-1;
		else if (cmp > 0)
			left = mid+1;
		else
			return (1);
	}
	return (0);
}
